stellar/txobserver/main.go

```diff
@@ -48,111 +59,149 @@ LedgerLoop:
+					// as withdrawn so it's not possible to cancel it.
+				} else {
+					otx.State = store.FailedState
+				}
+				err := o.store.UpsertOutgoingStellarTransaction(context.TODO(), otx)
+				if err != nil {
+					return errors.Wrapf(err, "error upserting outgoing transaction: %s", otx.Hash)
+				}
 			}
 
-			for _, tx := range txs.Embedded.Records {
-				otx, hashExists := outgoingBridgeTransactionsMap[tx.Hash]
+			if tx.InnerTransaction != nil {
+				otx, hashExists := outgoingBridgeTransactionsHashMap[tx.InnerTransaction.Hash]
 				if hashExists {
 					if tx.Successful {
 						otx.State = store.SuccessState
+						// TODO [important] update IncomingEthereumTransaction state
+						// as withdrawn so it's not possible to cancel it.
 					} else {
 						otx.State = store.FailedState
 					}
 					err := o.store.UpsertOutgoingStellarTransaction(context.TODO(), otx)
 					if err != nil {
-						o.log.WithFields(slog.F{
-							"error": err,
-							"hash":  otx.Hash,
-						}).Error("Error upserting outgoing transactions")
-						time.Sleep(time.Second)
-						continue LedgerLoop
+						return errors.Wrapf(err, "error upserting outgoing transaction: %s", otx.Hash)
 					}
 				}
+			}
 
-				if tx.InnerTransaction != nil {
-					otx, hashExists := outgoingBridgeTransactionsMap[tx.InnerTransaction.Hash]
-					if hashExists {
-						if tx.Successful {
-							otx.State = store.SuccessState
-						} else {
-							otx.State = store.FailedState
-						}
+			seqnums, sourceExists := outgoingBridgeTransactionsSourceMap[tx.Account]
+			if sourceExists {
+				for seqnum, otx := range seqnums {
+					if tx.AccountSequence >= seqnum && otx.Hash != tx.Hash &&
+						(tx.InnerTransaction == nil ||
+							(tx.InnerTransaction != nil && otx.Hash != tx.InnerTransaction.Hash)) {
+						// Account sequence number was used for another transaction or bump thus otx transaction
+						// is irrevocably invalid and can be marked as expired.
+						otx.State = store.ExpiredState
 						err := o.store.UpsertOutgoingStellarTransaction(context.TODO(), otx)
 						if err != nil {
-							o.log.WithFields(slog.F{
-								"error": err,
-								"hash":  otx.Hash,
-							}).Error("Error upserting outgoing transactions")
-							time.Sleep(time.Second)
-							continue LedgerLoop
+							return errors.Wrapf(err, "error upserting outgoing transaction: %s", otx.Hash)
 						}
 					}
 				}
-
-				cursor = tx.PagingToken()
 			}
-		}
 
-		// Mark all txs with expired time + buffer as expired.
-		expiredBefore := ledger.ClosedAt.Add(-time.Minute)
-		count, err := o.store.MarkOutgoingStellarTransactionExpired(context.TODO(), expiredBefore)
-		if err != nil {
-			o.log.WithField("error", err).Error("Error marking outgoing transactions as expired")
-			time.Sleep(time.Second)
-			continue LedgerLoop
+			cursor = tx.PagingToken()
 		}
+	}
+
+	// Mark all expired time + buffer and irrevocably invalid txs as expired.
+	expiredBefore := ledger.ClosedAt.Add(-time.Minute)
```

Comment:
Why is a buffer needed here? If the last known ledger closed at T=42 then, even if the validator is out of sync and late by hours or days or months, the "real" last ledger on Stellar closed at T'>=42. So any transaction that expires at T_exp<42 has irrevocably expired.

Reply:
Good call, fixed in 016667d. Probably leftover from the previous code using local time.